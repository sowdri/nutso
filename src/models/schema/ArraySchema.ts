import { OptionalFlag } from "../OptionalFlag";
import { Schema } from "./Schema";

/**
 * E => Element
 * T => Type (this is the array)
 * P => Parent of the array
 *
 * Look at this type, we need all 3 types
 */
export type ArraySchema<E, T extends E[], R, P = unknown> = {
  type: "array";
  optional?: OptionalFlag<R, P>;
  minItems?: number;
  maxItems?: number;
  items: Schema<E, R>; // TODO check how this works in schema definition
};
