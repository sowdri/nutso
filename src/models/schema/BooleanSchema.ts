import { OptionalFlag } from "../OptionalFlag";
import { ValidationFn } from "../ValidationFn";

export type BooleanSchema<R = boolean, P = unknown> = {
  type: "boolean";
  optional?: OptionalFlag<R, P>;
  validationFn?: ValidationFn<boolean, R, P>;
};
