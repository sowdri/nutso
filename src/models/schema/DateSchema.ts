import { OptionalFlag } from "../OptionalFlag";
import { ValidationFn } from "../ValidationFn";

export type DateSchema<R = Date, P = unknown> = {
  type: "date";
  optional?: OptionalFlag<R, P>;
  validationFn?: ValidationFn<Date, R, P>;
};
