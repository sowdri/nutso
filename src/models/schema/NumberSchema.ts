import { OptionalFlag } from "../OptionalFlag";
import { ValidationFn } from "../ValidationFn";

export type NumberSchema<R = number, P = unknown> = {
  type: "number";
  min?: number;
  max?: number;
  pattern?: RegExp;
  optional?: OptionalFlag<R, P>;
  validationFn?: ValidationFn<number, R, P>;
};
