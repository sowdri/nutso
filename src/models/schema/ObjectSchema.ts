import { Schema } from "./Schema";
import { OptionalFlag } from "../OptionalFlag";

export type ObjectSchema<T, R = T, P = unknown> = {
  type: "object";
  optional?: OptionalFlag<R, P>;
  properties: {
    [K in keyof T]: Schema<T[K], R, T>;
  };
};
