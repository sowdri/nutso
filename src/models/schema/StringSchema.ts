import { OptionalFlag } from "../OptionalFlag";
import { ValidationFn } from "../ValidationFn";
import { Schema } from "./Schema";

/*
type ValidatorFnResult1 = {
  errorMessage: string;
};

type ValidationFn1<T, R, P> = (args: { value: T; parent: P; root: R }) => ValidatorFnResult1 | void;

*/
export type StringSchema<R = string, P = unknown> = {
  type: "string";
  optional?: OptionalFlag<R, P>;
  minLength?: number;
  maxLength?: number;
  pattern?: RegExp;
  validationFn?: ValidationFn<string, R, P>;
};

const nakedSchema: Schema<string> = {
  type: "string",
  validationFn: (args) => {},
};

type Customer = {
  name: string;
};

const schema: Schema<Customer> = {
  type: "object",
  properties: {
    name: {
      type: "string",
      validationFn: (args) => {},
    },
  },
};

type Bar = {
  bar: string;
};

type Foo = {
  bar: Bar;
};

const fooSchema: Schema<Foo> = {
  type: "object",
  properties: {
    bar: {
      type: "object",
      properties: {
        bar: {
          type: "string",
          validationFn: (args) => {},
        },
      },
    },
  },
};
