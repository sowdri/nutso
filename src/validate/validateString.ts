import { StringResult } from "../models/result/StringResult";
import { StringSchema } from "../models/schema/StringSchema";
import { isNil, isString } from "../utils/typeChecker";
import { validationFnExecutor } from "../utils/validationFnExecutor";
import { optionalFlagValidator, isOptional } from "../utils/optionalFlagValidator";

export const validateString = <R, P>(args: {
  value: any;
  root: R;
  parent: P;
  schema: StringSchema<R, P>;
}): StringResult => {
  const { value, schema } = args;
  //

  // isnil
  if (isNil(value)) {
    return optionalFlagValidator({ ...args, flag: schema.optional });
  }

  // check if type is string
  if (!isString(value)) {
    return {
      isValid: false,
      errorMessage: `Should be a string.`,
      errorPath: [],
    };
  }

  const str = value as string;

  // check if empty
  if (str === "") {
    const optional = isOptional({ ...args, flag: schema.optional });
    if (!optional)
      return {
        isValid: false,
        errorMessage: `Should not be empty.`,
        errorPath: [],
      };
  }

  // min length
  if (!isNil(schema.minLength) && str.length < schema.minLength!) {
    return {
      isValid: false,
      errorMessage: `Should be at least ${schema.minLength} characters.`,
      errorPath: [],
    };
  }

  // max length
  if (!isNil(schema.maxLength) && str.length > schema.maxLength!) {
    return {
      isValid: false,
      errorMessage: `Should not be longer than ${schema.maxLength} characters.`,
      errorPath: [],
    };
  }

  // pattern
  if (!isNil(schema.pattern)) {
    const match = schema.pattern!.test(value);
    if (!match) {
      return {
        isValid: false,
        errorMessage: `Should match the pattern ${schema.pattern} .`,
        errorPath: [],
      };
    }
  }

  // validationFn
  if (schema.validationFn) {
    const result = validationFnExecutor({ ...args, value: str, validationFn: schema.validationFn });
    if (result) return result;
  }

  return {
    isValid: true,
    errorMessage: ``,
    errorPath: [],
  };
};
