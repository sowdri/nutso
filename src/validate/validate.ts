import { Result } from "../models/result/Result";
import { BooleanSchema } from "../models/schema/BooleanSchema";
import { DateSchema } from "../models/schema/DateSchema";
import { NumberSchema } from "../models/schema/NumberSchema";
import { Schema } from "../models/schema/Schema";
import { StringSchema } from "../models/schema/StringSchema";
import { validateArray } from "./validateArray";
import { validateBoolean } from "./validateBoolean";
import { validateDate } from "./validateDate";
import { validateNumber } from "./validateNumber";
import { validateObject } from "./validateObject";
import { validateString } from "./validateString";

/**
 * This is an internal API that takes the root as an argument
 * and called recursively from `validateObject` and `validateArray` methods
 *
 * @param value
 * @param parent
 * @param schema
 */
export const _validate = <T, P>(args: { value: T | null; parent: P; schema: Schema<T, P> }): Result<T> => {
  //

  if (!args.schema) {
    throw new Error("Schema should not be null");
  }

  /**
   * validate object against the schema
   */
  switch (args.schema.type) {
    case "number":
      return validateNumber({ ...args, schema: args.schema as NumberSchema<P> }) as Result<T>;
    case "string":
      return validateString({ ...args, schema: args.schema as StringSchema<P> }) as Result<T>;
    case "boolean":
      return validateBoolean({ ...args, schema: args.schema as BooleanSchema<P> }) as Result<T>;
    case "array":
      return validateArray({ value: args.value as any, parent: args.parent, schema: args.schema as any }) as Result<T>;
    case "object":
      return validateObject({ ...args, schema: args.schema }) as Result<T>;
    case "date":
      return validateDate({ ...args, schema: args.schema as DateSchema<P> }) as Result<T>;
  }
  throw new Error(`Unhandled data type`);
};

// public api
export const validate = <T>(value: T, schema: Schema<T>): Result<T> => {
  return _validate({ value, parent: null as any, schema });
};
