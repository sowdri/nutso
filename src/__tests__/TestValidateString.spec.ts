import { StringSchema } from "../models/schema/StringSchema";
import { validate } from "../validate/validate";
import { validateString } from "../validate/validateString";

test(`Basic`, () => {
  const str = "foo";
  const schema: StringSchema = {
    type: "string",
  };
  const result1 = validateString({ value: str, root: str, schema, parent: undefined as any });
  expect(result1.isValid).toBe(true);

  const result2 = validate(str, schema);
  expect(result2.isValid).toBe(true);
});

test(`Empty string - invalid`, () => {
  const str = "";
  const schema: StringSchema = {
    type: "string",
  };
  const result1 = validateString({ value: str, root: str, schema, parent: undefined as any });
  expect(result1.isValid).toBe(false);

  const result2 = validate(str, schema);
  expect(result2.isValid).toBe(false);
});

test(`Empty string - not optional`, () => {
  const str = "";
  const schema: StringSchema = {
    type: "string",
    optional: () => false,
  };
  const result1 = validateString({ value: str, root: str, schema, parent: undefined as any });
  expect(result1.isValid).toBe(false);

  const result2 = validate(str, schema);
  expect(result2.isValid).toBe(false);
});

test(`Basic - invalid min-length`, () => {
  const str = "";
  const schema: StringSchema = {
    type: "string",
    minLength: 3,
  };
  const result1 = validateString({ value: str, root: str, schema, parent: undefined as any });
  expect(result1.isValid).toBe(false);

  const result2 = validate(str, schema);
  expect(result2.isValid).toBe(false);
});

test(`Basic - invalid max-length`, () => {
  const str = "foo-bar";
  const schema: StringSchema = {
    type: "string",
    maxLength: 3,
  };
  const result1 = validateString({ value: str, root: str, schema, parent: undefined as any });
  expect(result1.isValid).toBe(false);

  const result2 = validate(str, schema);
  expect(result2.isValid).toBe(false);
});

test(`Regex match`, () => {
  const schema: StringSchema = {
    type: "string",
    pattern: /5\d{3}/,
  };
  const valid = "5085";
  const invalid = "508";

  const result1 = validateString({ value: valid, root: valid, schema, parent: undefined as any });
  expect(result1.isValid).toBe(true);

  const result2 = validate(invalid, schema);
  expect(result2.isValid).toBe(false);
});

test(`Validation function - check value`, () => {
  const schema: StringSchema = {
    type: "string",
    validationFn: (args) => {
      if (args.value === "5085") return;
      return {
        errorMessage: "Invalid",
      };
    },
  };
  const valid = "5085";

  const result = validateString({ value: valid, root: valid, schema, parent: undefined as any });
  expect(result).toMatchInlineSnapshot(`
{
  "errorMessage": "",
  "errorPath": [],
  "isValid": true,
}
`);
});

test(`Validation function - return validation result`, () => {
  const schema: StringSchema = {
    type: "string",
    validationFn: (value) => {
      return {
        errorMessage: "Custom validation failed",
      };
    },
  };
  const valid = "5085";

  const result = validateString({ value: valid, root: valid, schema, parent: undefined as any });
  expect(result).toMatchSnapshot();
});

test(`Validation function - throw error`, () => {
  const schema: StringSchema = {
    type: "string",
    validationFn: (value) => {
      throw new Error(`Validation fn threw!`);
    },
  };
  const valid = "5085";

  const result = validateString({ value: valid, root: valid, schema, parent: undefined as any });
  expect(result).toMatchSnapshot();
});

test(`Validation function - throw error object - valid`, () => {
  const schema: StringSchema = {
    type: "string",
    validationFn: (value) => {
      throw { message: "Custom error object, with message field" };
    },
  };
  const valid = "5085";

  const result = validateString({ value: valid, root: valid, schema, parent: undefined as any });
  expect(result).toMatchSnapshot();
});

test(`Validation function - throw error object - invalid`, () => {
  const schema: StringSchema = {
    type: "string",
    validationFn: (value) => {
      throw { foo: "Custom error object, with message field" };
    },
  };
  const valid = "5085";

  const result = validateString({ value: valid, root: valid, schema, parent: undefined as any });
  expect(result).toMatchSnapshot();
});
