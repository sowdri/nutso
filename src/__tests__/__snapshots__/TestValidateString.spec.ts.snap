// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validation function - throw error 1`] = `
{
  "errorMessage": "Validation fn threw!",
  "errorPath": [],
  "isValid": false,
}
`;

exports[`Validation function - throw error object - invalid 1`] = `
{
  "errorMessage": "Exception in validationFn",
  "errorPath": [],
  "isValid": false,
}
`;

exports[`Validation function - throw error object - valid 1`] = `
{
  "errorMessage": "Custom error object, with message field",
  "errorPath": [],
  "isValid": false,
}
`;

exports[`Validation function 1`] = `
{
  "errorMessage": "Custom validation failed",
  "errorPath": [],
  "isValid": false,
}
`;
